pkg/threescale/backend/manager.go

```diff
@@ -0,0 +1,370 @@
+		newlyCachedApplication, err := b.handleCacheMiss(request, cacheKey)
+		if err != nil {
+			return app, cacheKey, err
+		}
+		app = newlyCachedApplication
+	}
+	return app, cacheKey, nil
+}
+
+func (b *Backend) generateCacheKeyFromRequest(apiCall threescale.Request, transaction api.Transaction) string {
+	svcID := apiCall.Service
+	appID := b.getAppIdFromRequest(transaction)
+	return fmt.Sprintf("%s_%s", svcID, appID)
+}
+
+// getAppIdFromRequest prioritizes and returns a user key if present, defaulting to app id otherwise
+func (b *Backend) getAppIdFromRequest(transaction api.Transaction) string {
+	appID := transaction.Params.UserKey
+
+	if appID != "" {
+		return appID
+	}
+	return transaction.Params.AppID
+}
+
+// handleCacheMiss will attempt to call remote 3scale with a blank request and the required
+// extensions enabled to learn current state for caching purposes.
+// Sets the value in the cache and returns the newly cached value for re-use if required
+func (b *Backend) handleCacheMiss(request threescale.Request, cacheKey string) (Application, error) {
+	var app Application
+	emptyTransaction := b.blankRequestFromTransactions(request.Transactions)
+
+	emptyRequest := threescale.Request{
+		Auth:       request.Auth,
+		Extensions: api.Extensions{api.HierarchyExtension: "1"},
+		Service:    request.Service,
+		Transactions: []api.Transaction{
+			emptyTransaction,
+		},
+	}
+	resp, err := b.remoteAuth(emptyRequest)
+	if err != nil {
+		return app, err
+	}
+	app = b.applicationFromResponse(resp)
+
+	b.cache.Set(cacheKey, app)
+	return app, nil
+}
+
+// blankRequestFromTransactions is used when we need to gather state from the remote backend
+// it parses the transactions for their auth and augments the request to provide nil metrics
+func (b *Backend) blankRequestFromTransactions(transactions []api.Transaction) api.Transaction {
+	transactionCopy := make([]api.Transaction, len(transactions))
+	copy(transactionCopy, transactions)
+	transactionCopy[0].Metrics = nil
+
+	return transactionCopy[0]
+}
+
+// applicationFromResponse builds an Application from a response from 3scale client
+func (b *Backend) applicationFromResponse(resp *threescale.AuthorizeResult) Application {
+	limitCounter := make(map[string]map[api.Period]*Limit)
+
+	for metric, report := range resp.UsageReports {
+		l := &Limit{
+			current: report.CurrentValue,
+			max:     report.MaxValue,
+		}
+		limitCounter[metric] = make(map[api.Period]*Limit)
+		limitCounter[metric][report.PeriodWindow.Period] = l
+	}
+
+	return Application{
+		LastResponse:     resp,
+		LimitCounter:     limitCounter,
+		UnlimitedCounter: make(map[string]int),
+	}
+}
+
+// computeAffectedMetrics takes the provided metrics and hierarchy, and uses the hierarchy
+// to determine how the metrics are affected, modifying the metrics accordingly
+func (b *Backend) computeAffectedMetrics(hierarchy api.Hierarchy, metrics api.Metrics) api.Metrics {
```

Comment:
perhaps we can move this logic to `api.Hierarchy` to have a method like `func (h *Hierarchy) ComputeAffectedMetrics(metrics api.Metrics) api.Metrics`? Or conversely add it to `api.Metrics` and receive an `api.Hierarchy`?

Reply:
sure, i think you have pointed out a similar idea elsewhere. how about getting it working in place here and i can move package specific logic out at a later date if that sounds ok? wdyt

Reply:
done in https://github.com/3scale/3scale-go-client/pull/32, will integrate